File: src/store/modules/comments.js
import apiService from '@/services/api';

export default {
  namespaced: true,
  
  state: {
    comments: {},  // { propertyId: [comment1, comment2, ...] }
    currentComment: null,
    replies: [],
    loading: false,
    error: null,
    pagination: {
      total: 0,
      currentPage: 1,
      totalPages: 1,
    }
  },
  
  getters: {
    isLoading: state => state.loading,
    error: state => state.error,
    getComments: state => propertyId => state.comments[propertyId] || [],
    getPropertyRating: state => propertyId => {
      const comments = state.comments[propertyId] || [];
      if (comments.length === 0) return 0;
      
      const sum = comments.reduce((total, comment) => total + comment.rating, 0);
      return (sum / comments.length).toFixed(1);
    },
    commentCount: state => propertyId => (state.comments[propertyId] || []).length,
    allReplies: state => state.replies
  },
  
  mutations: {
    SET_LOADING(state, status) {
      state.loading = status;
    },
    
    SET_ERROR(state, error) {
      state.error = error;
    },
    
    SET_COMMENTS(state, { propertyId, comments }) {
      state.comments = {
        ...state.comments,
        [propertyId]: comments
      };
    },
    
    SET_CURRENT_COMMENT(state, comment) {
      state.currentComment = comment;
    },
    
    SET_REPLIES(state, replies) {
      state.replies = replies;
    },
    
    SET_PAGINATION(state, pagination) {
      state.pagination = pagination;
    },
    
    ADD_COMMENT(state, { propertyId, comment }) {
      if (!state.comments[propertyId]) {
        state.comments[propertyId] = [];
      }
      state.comments[propertyId] = [comment, ...state.comments[propertyId]];
    },
    
    UPDATE_COMMENT(state, { propertyId, commentId, data }) {
      const comments = state.comments[propertyId] || [];
      const index = comments.findIndex(c => c.id === commentId);
      
      if (index !== -1) {
        state.comments[propertyId][index] = {
          ...state.comments[propertyId][index],
          ...data
        };
      }
    },
    
    DELETE_COMMENT(state, { propertyId, commentId }) {
      if (state.comments[propertyId]) {
        state.comments[propertyId] = state.comments[propertyId].filter(
          c => c.id !== commentId
        );
      }
    },
    
    ADD_REPLY(state, reply) {
      state.replies = [...state.replies, reply];
      
      // 更新評論的回覆計數
      const propertyId = state.currentComment?.property_id;
      const commentId = reply.comment_id;
      
      if (propertyId && state.comments[propertyId]) {
        const commentIndex = state.comments[propertyId].findIndex(c => c.id === commentId);
        if (commentIndex !== -1) {
          state.comments[propertyId][commentIndex].replyCount += 1;
        }
      }
    },
    
    UPDATE_REPLY(state, { replyId, data }) {
      const index = state.replies.findIndex(r => r.id === replyId);
      if (index !== -1) {
        state.replies[index] = { ...state.replies[index], ...data };
      }
    },
    
    DELETE_REPLY(state, { replyId, commentId, propertyId }) {
      // 移除回覆
      state.replies = state.replies.filter(r => r.id !== replyId);
      
      // 更新評論的回覆計數
      if (propertyId && commentId && state.comments[propertyId]) {
        const commentIndex = state.comments[propertyId].findIndex(c => c.id === commentId);
        if (commentIndex !== -1 && state.comments[propertyId][commentIndex].replyCount > 0) {
          state.comments[propertyId][commentIndex].replyCount -= 1;
        }
      }
    },
  },
  
  actions: {
    // 獲取特定房源的評論
    async fetchComments({ commit }, { propertyId, page = 1, perPage = 20 }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.comments.getPropertyComments(propertyId, { page, perPage });
        
        commit('SET_COMMENTS', { propertyId, comments: response.comments });
        commit('SET_PAGINATION', {
          total: response.total,
          currentPage: response.current_page,
          totalPages: response.pages
        });
        
        return response.comments;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法獲取評論');
        console.error('獲取評論失敗:', error);
        return [];
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 新增評論
    async addComment({ commit }, { propertyId, content, rating }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.comments.createComment(propertyId, { content, rating });
        commit('ADD_COMMENT', { propertyId, comment: response });
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法新增評論');
        console.error('新增評論失敗:', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 更新評論
    async updateComment({ commit }, { propertyId, commentId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.comments.updateComment(commentId, data);
        commit('UPDATE_COMMENT', { propertyId, commentId, data: response });
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法更新評論');
        console.error('更新評論失敗:', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 刪除評論
    async deleteComment({ commit }, { propertyId, commentId }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        await apiService.comments.deleteComment(commentId);
        commit('DELETE_COMMENT', { propertyId, commentId });
        return true;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法刪除評論');
        console.error('刪除評論失敗:', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 獲取評論回覆
    async fetchReplies({ commit }, { commentId, page = 1, perPage = 50 }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.comments.getReplies(commentId, { page, perPage });
        
        commit('SET_CURRENT_COMMENT', response.comment);
        commit('SET_REPLIES', response.replies);
        commit('SET_PAGINATION', {
          total: response.total,
          currentPage: response.current_page,
          totalPages: response.pages
        });
        
        return response.replies;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法獲取回覆');
        console.error('獲取回覆失敗:', error);
        return [];
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 新增回覆
    async addReply({ commit, state }, { commentId, content }) {
      try {
        const response = await apiService.comments.createReply(commentId, { content });
        
        // 找出評論所屬的房源 ID
        let propertyId = null;
        for (const [propId, commentsList] of Object.entries(state.comments)) {
          const comment = commentsList.find(c => c.id === commentId);
          if (comment) {
            propertyId = parseInt(propId);
            break;
          }
        }
        
        // 新增回覆到 state
        commit('ADD_REPLY', response);
        
        // 更新評論的回覆計數
        if (propertyId) {
          const commentsList = state.comments[propertyId];
          const commentIndex = commentsList.findIndex(c => c.id === commentId);
          
          if (commentIndex !== -1) {
            const updatedReplyCount = (commentsList[commentIndex].replyCount || 0) + 1;
            
            commit('UPDATE_COMMENT', {
              propertyId,
              commentId,
              data: {
                replyCount: updatedReplyCount
              }
            });
          }
        }
        
        return response;
      } catch (error) {
        console.error('新增回覆失敗:', error);
        throw error;
      }
    },
    
    // 更新回覆
    async updateReply({ commit }, { replyId, content }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.comments.updateReply(replyId, { content });
        commit('UPDATE_REPLY', { replyId, data: response });
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法更新回覆');
        console.error('更新回覆失敗:', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 刪除回覆
    async deleteReply({ commit, state }, replyId) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        // 在刪除前先獲取回覆相關信息
        const replyToDelete = state.replies.find(r => r.id === replyId);
        if (!replyToDelete) {
          throw new Error('找不到要刪除的回覆');
        }
        
        const commentId = replyToDelete.comment_id;
        
        // 找出評論所屬的房源 ID
        let propertyId = null;
        if (state.currentComment && state.currentComment.id === commentId) {
          propertyId = state.currentComment.property_id;
        } else {
          // 查找所有房源的評論
          for (const [propId, commentsList] of Object.entries(state.comments)) {
            const comment = commentsList.find(c => c.id === commentId);
            if (comment) {
              propertyId = parseInt(propId);
              break;
            }
          }
        }
        
        // 調用 API 刪除回覆
        await apiService.comments.deleteReply(replyId);
        
        // 同時更新 state 中的回覆和評論計數
        commit('DELETE_REPLY', { replyId, commentId, propertyId });
        
        return true;
      } catch (error) {
        commit('SET_ERROR', error.message || '無法刪除回覆');
        console.error('刪除回覆失敗:', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 點讚/取消點讚評論
    async likeComment({ commit, state }, commentId) {
      try {
        const response = await apiService.comments.likeComment(commentId);
        
        // 找出評論所屬的房源 ID
        let propertyId = null;
        for (const [propId, commentsList] of Object.entries(state.comments)) {
          const comment = commentsList.find(c => c.id === commentId);
          if (comment) {
            propertyId = parseInt(propId);
            break;
          }
        }
        
        if (propertyId) {
          // 更新評論的讚數和讚列表
          commit('UPDATE_COMMENT', {
            propertyId,
            commentId,
            data: {
              likes: response.likes,
              likedBy: response.likedBy
            }
          });
        }
        
        return response;
      } catch (error) {
        console.error('點讚評論失敗:', error);
        throw error;
      }
    },
    
    // 點讚/取消點讚回覆
    async likeReply({ commit }, replyId) {
      try {
        const response = await apiService.comments.likeReply(replyId);
        
        commit('UPDATE_REPLY', {
          replyId,
          data: {
            like_count: response.like_count,
            likes: response.likes
          }
        });
        
        return response;
      } catch (error) {
        console.error('點讚回覆失敗:', error);
        throw error;
      }
    }
  }
};

File: backend/app/api/comments.py
from flask import Blueprint, request, jsonify, session
from app.models.comments import Comment, Reply, CommentLike, ReplyLike, Report
from app.extensions import db
from sqlalchemy.exc import SQLAlchemyError # type: ignore
from datetime import datetime
from functools import wraps
from app.api import comments_bp

# 用戶身份驗證裝飾器
def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'user_id' not in session:
            return jsonify({"message": "請先登入"}), 401
        
        return f(*args, **kwargs)
    return decorated

# 評論相關 API
@comments_bp.route('/property/<int:property_id>', methods=['GET'])
def get_property_comments(property_id):
    """獲取特定房源的所有評論"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 20, type=int)
    
    query = Comment.query.filter_by(property_id=property_id).order_by(Comment.created_at.desc())
    comments_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'success': True,
        'comments': [comment.to_dict() for comment in comments_pagination.items],
        'total': comments_pagination.total,
        'pages': comments_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)

@comments_bp.route('/property/<int:property_id>', methods=['POST'])
@login_required
def create_comment(property_id):
    """對特定房源發表評論"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    if not data.get('content'):
        return jsonify({
            'success': False,
            'error': '評論內容不能為空'
        }), 400
    
    if not data.get('rating') or not isinstance(data.get('rating'), int) or data.get('rating') < 1 or data.get('rating') > 5:
        return jsonify({
            'success': False,
            'error': '評分必須為1-5的整數'
        }), 400
    
    try:
        comment = Comment(
            property_id=property_id,
            user_id=user_id,
            content=data.get('content'),
            rating=data.get('rating')
        )
        
        db.session.add(comment)
        db.session.commit()
        
        return jsonify(comment.to_dict()), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'發表評論失敗: {str(e)}'}), 500

@comments_bp.route('/<int:comment_id>', methods=['PUT'])
@login_required
def update_comment(comment_id):
    """更新評論內容和評分"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    comment = Comment.query.get_or_404(comment_id)
    
    # 只有作者可以更新評論
    if comment.user_id != user_id:
        return jsonify({
            'success': False,  
            'error': '您沒有權限更新此評論'
        }), 403
    
    if not data.get('content'):
        return jsonify({
            'success': False,  
            'error': '評論內容不能為空'
        }), 400
    
    if 'rating' in data and (not isinstance(data.get('rating'), int) or data.get('rating') < 1 or data.get('rating') > 5):
        return jsonify({
            'success': False,  
            'error': '評分必須為1-5的整數'
        }), 400
    
    try:
        comment.content = data.get('content', comment.content)
        if 'rating' in data:
            comment.rating = data.get('rating')
        
        db.session.commit()
        
        comment_dict = comment.to_dict()
        return jsonify({
            'success': True,
            'message': '評論更新成功',
            'comment': comment_dict
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,  
            'error': f'更新評論失敗: {str(e)}'
        }), 500

@comments_bp.route('/<int:comment_id>', methods=['DELETE'])
@login_required
def delete_comment(comment_id):
    """刪除評論 (僅限作者或管理員)"""
    user_id = session.get('user_id')
    
    comment = Comment.query.get_or_404(comment_id)
    
    # 只有作者或管理員可以刪除評論
    if comment.user_id != user_id and not session.get('is_admin', False):
        return jsonify({
            'success': False,  
            'error': '您沒有權限刪除此評論'
        }), 403
    
    try:
        db.session.delete(comment)
        db.session.commit()
        return jsonify({
            'success': True,  
            'message': '評論已成功刪除'
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,  
            'error': f'刪除評論失敗: {str(e)}'
        }), 500

# 回覆相關 API
@comments_bp.route('/<int:comment_id>/replies', methods=['GET'])
def get_replies(comment_id):
    """獲取特定評論的所有回覆"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 50, type=int)
    
    # 確認評論存在
    comment = Comment.query.get_or_404(comment_id)
    
    query = Reply.query.filter_by(comment_id=comment_id).order_by(Reply.created_at.asc())
    replies_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'comment': comment.to_dict(),
        'replies': [reply.to_dict() for reply in replies_pagination.items],
        'total': replies_pagination.total,
        'pages': replies_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)

@comments_bp.route('/<int:comment_id>/replies', methods=['POST'])
@login_required
def create_reply(comment_id):
    """對特定評論發表回覆"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    # 確認評論存在
    comment = Comment.query.get_or_404(comment_id)
    
    if not data.get('content'):
        return jsonify({'error': '回覆內容不能為空'}), 400
    
    try:
        reply = Reply(
            comment_id=comment_id,
            user_id=user_id,
            content=data.get('content')
        )
        
        db.session.add(reply)
        db.session.commit()
        
        return jsonify(reply.to_dict()), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'發表回覆失敗: {str(e)}'}), 500

@comments_bp.route('/replies/<int:reply_id>', methods=['PUT'])
@login_required
def update_reply(reply_id):
    """更新回覆內容"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    reply = Reply.query.get_or_404(reply_id)
    
    # 只有作者可以更新回覆
    if reply.user_id != user_id:
        return jsonify({
            'success': False,  
            'error': '您沒有權限更新此回覆'
        }), 403
    
    if not data.get('content'):
        return jsonify({
            'success': False,  
            'error': '回覆內容不能為空'
        }), 400
    
    try:
        reply.content = data.get('content')
        db.session.commit()
        
        reply_dict = reply.to_dict()
        return jsonify({
            'success': True,
            'message': '回覆更新成功',
            'reply': reply_dict
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,  
            'error': f'更新回覆失敗: {str(e)}'
        }), 500

@comments_bp.route('/replies/<int:reply_id>', methods=['DELETE'])
@login_required
def delete_reply(reply_id):
    """刪除回覆 (僅限作者或管理員)"""
    user_id = session.get('user_id')
    
    reply = Reply.query.get_or_404(reply_id)
    
    # 只有作者或管理員可以刪除回覆
    if reply.user_id != user_id and not session.get('is_admin', False):
        return jsonify({
            'success': False,  
            'error': '您沒有權限刪除此回覆'
        }), 403
    
    try:
        db.session.delete(reply)
        db.session.commit()
        return jsonify({
            'success': True,  
            'message': '回覆已成功刪除'
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,  
            'error': f'刪除回覆失敗: {str(e)}'
        }), 500

@comments_bp.route('/<int:comment_id>/like', methods=['POST'])
@login_required
def like_comment(comment_id):
    """對評論按讚/取消讚"""
    user_id = session.get('user_id')
    
    # 確認評論存在
    comment = Comment.query.get_or_404(comment_id)
    
    # 檢查是否已按讚
    existing_like = CommentLike.query.filter_by(comment_id=comment_id, user_id=user_id).first()
    is_liked = False
    
    try:
        if existing_like:
            # 取消讚
            db.session.delete(existing_like)
            message = "已取消對評論的讚"
        else:
            # 按讚
            like = CommentLike(comment_id=comment_id, user_id=user_id)
            db.session.add(like)
            message = "已成功對評論按讚"
            is_liked = True
            
        db.session.commit()
        
        # 重新查詢以獲取最新的點讚資訊
        likes_count = CommentLike.query.filter_by(comment_id=comment_id).count()
        liked_by = [like.user_id for like in CommentLike.query.filter_by(comment_id=comment_id).all()]
        
        return jsonify({
            'success': True,  
            'message': message, 
            'likes': likes_count,
            'likedBy': liked_by,
            'isLiked': is_liked
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,  
            'error': f'操作失敗: {str(e)}'
        }), 500

@comments_bp.route('/replies/<int:reply_id>/like', methods=['POST'])
@login_required
def like_reply(reply_id):
    """對回覆按讚/取消讚"""
    user_id = session.get('user_id')
    
    # 確認回覆存在
    reply = Reply.query.get_or_404(reply_id)
    
    # 檢查是否已按讚
    existing_like = ReplyLike.query.filter_by(reply_id=reply_id, user_id=user_id).first()
    is_liked = False
    
    try:
        if existing_like:
            # 取消讚
            db.session.delete(existing_like)
            message = "已取消對回覆的讚"
        else:
            # 按讚
            like = ReplyLike(reply_id=reply_id, user_id=user_id)
            db.session.add(like)
            message = "已成功對回覆按讚"
            is_liked = True
            
        db.session.commit()
        
        # 重新查詢以獲取最新的點讚資訊
        likes_count = ReplyLike.query.filter_by(reply_id=reply_id).count()
        liked_by = [like.user_id for like in ReplyLike.query.filter_by(reply_id=reply_id).all()]
        
        return jsonify({
            'success': True, 
            'message': message, 
            'like_count': likes_count,
            'likes': liked_by,
            'isLiked': is_liked
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False, 
            'error': f'操作失敗: {str(e)}'
        }), 500

@comments_bp.route('/report', methods=['POST'])
@login_required
def report_content():
    """舉報評論或回覆"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    content_type = data.get('contentType')
    content_id = data.get('contentId')
    reasons = data.get('reasons', [])
    description = data.get('description', '')
    
    # 驗證參數
    if not content_type or not content_id:
        return jsonify({'error': '缺少必要參數'}), 400
    
    if content_type not in ['comment', 'reply']:
        return jsonify({'error': '無效的內容類型'}), 400
    
    if not reasons or not isinstance(reasons, list) or len(reasons) == 0:
        return jsonify({'error': '至少需要選擇一個舉報原因'}), 400
    
    # 確認內容存在
    if content_type == 'comment':
        content = Comment.query.get_or_404(content_id)
    else:  # reply
        content = Reply.query.get_or_404(content_id)
    
    # 確認用戶沒有舉報過同一內容
    existing_report = Report.query.filter_by(
        reporter_id=user_id,
        content_type=content_type,
        content_id=content_id
    ).first()
    
    if existing_report:
        return jsonify({
            'success': False,
            'message': '您已經舉報過此內容',
            'report': existing_report.to_dict()
        }), 400
    
    try:
        # 創建舉報記錄
        report = Report(
            reporter_id=user_id,
            content_type=content_type,
            content_id=content_id,
            reasons=','.join(reasons),
            description=description
        )
        
        db.session.add(report)
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': '舉報已提交，感謝您幫助我們維護社區環境',
            'report': report.to_dict()
        }), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({
            'success': False,
            'error': f'舉報提交失敗: {str(e)}'
        }), 500

@comments_bp.route('/report/<int:report_id>', methods=['GET'])
@login_required
def get_report(report_id):
    """獲取舉報詳情"""
    user_id = session.get('user_id')
    is_admin = session.get('is_admin', False)
    
    report = Report.query.get_or_404(report_id)
    
    # 只有舉報者本人或管理員可以查看舉報詳情
    if report.reporter_id != user_id and not is_admin:
        return jsonify({'error': '您沒有權限查看此舉報'}), 403
    
    return jsonify({
        'success': True,
        'report': report.to_dict()
    })

@comments_bp.route('/reports', methods=['GET'])
@login_required
def get_user_reports():
    """獲取用戶自己的舉報列表"""
    user_id = session.get('user_id')
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 20, type=int)
    
    query = Report.query.filter_by(reporter_id=user_id).order_by(Report.created_at.desc())
    reports_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'success': True,
        'reports': [report.to_dict() for report in reports_pagination.items],
        'total': reports_pagination.total,
        'pages': reports_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)